src/CVRF/cvrf_priv.c

```diff
@@ -0,0 +1,1518 @@
+
+		if (!xmlStrcmp(xmlTextReaderConstLocalName(reader), TAG_TRACKING_IDENTIFICATION)) {
+			xmlTextReaderNextElement(reader);
+			if (!xmlStrcmp(xmlTextReaderConstLocalName(reader), TAG_TRACKING_ID)) {
+				tracking->tracking_id = (char *)oscap_element_string_copy(reader);
+			}
+			if (!xmlStrcmp(xmlTextReaderConstLocalName(reader), TAG_TRACKING_ALIAS)) {
+				tracking->tracking_alias = (char *)oscap_element_string_copy(reader);
+			}
+		} else if (!xmlStrcmp(xmlTextReaderConstLocalName(reader), TAG_TRACKING_STATUS)) {
+			tracking->tracking_status = cvrf_doc_status_type_parse(reader);
+		} else if (!xmlStrcmp(xmlTextReaderConstLocalName(reader), TAG_TRACKING_VERSION)) {
+			tracking->tracking_version = (int)oscap_element_string_copy(reader);
+		} else if (!xmlStrcmp(xmlTextReaderConstLocalName(reader), TAG_TRACKING_INIT_RELEASE)) {
+			tracking->init_release_date = (char *)oscap_element_string_copy(reader);
+		} else if (!xmlStrcmp(xmlTextReaderConstLocalName(reader), TAG_TRACKING_CUR_RELEASE)) {
+			tracking->cur_release_date = (char *)oscap_element_string_copy(reader);
+		} else if (!xmlStrcmp(xmlTextReaderConstLocalName(reader), TAG_GENERATOR)) {
+			xmlTextReaderNextElement(reader);
+			if (!xmlStrcmp(xmlTextReaderConstLocalName(reader), TAG_GENERATOR_ENGINE)) {
+				tracking->generator_engine = (char *)oscap_element_string_copy(reader);
+				xmlTextReaderNextElement(reader);
+			}
+			if (!xmlStrcmp(xmlTextReaderConstLocalName(reader), TAG_GENERATOR_DATE)) {
+				tracking->generator_date = (char *)oscap_element_string_copy(reader);
+			}
+		}
+
+		xmlTextReaderNextNode(reader);
+	}
+
+	return tracking;
+}
+
+struct cvrf_product_tree *cvrf_product_tree_parse(xmlTextReaderPtr reader) {
+	__attribute__nonnull__(reader);
+
+	struct cvrf_product_tree *tree;
+	struct cvrf_branch *branch;
+	struct cvrf_relationship *relation;
+	struct cvrf_product_name *full_name;
+
+	/* allocate platform structure here */
+	tree = cvrf_product_tree_new();
+	if (tree == NULL)
+		return NULL;
+
+	/* If <empty /> then return, because there is no child element */
+	if (xmlTextReaderIsEmptyElement(reader))
+		return tree;
+
+	/* skip from <entry> node to next one */
+	xmlTextReaderNextElement(reader);
+
+	while (xmlStrcmp(xmlTextReaderConstLocalName(reader), TAG_CVRF_PRODUCT_TREE) != 0) {
+
+		if (xmlTextReaderNodeType(reader) != XML_READER_TYPE_ELEMENT) {
+			xmlTextReaderNextNode(reader);
+			continue;
+		}
+
+		if (!xmlStrcmp(xmlTextReaderConstLocalName(reader), TAG_CVRF_PRODUCT_NAME)) {
+			full_name = cvrf_product_name_new();
+			full_name->product_id = (char *)xmlTextReaderGetAttribute(reader, ATTR_PRODUCT_ID);
+			full_name->cpe = (char *)oscap_element_string_copy(reader);
+			tree->full_name = full_name;
+		} else if (!xmlStrcmp(xmlTextReaderConstLocalName(reader), TAG_BRANCH)) {
+			while(xmlStrcmp(xmlTextReaderConstLocalName(reader), TAG_BRANCH) == 0) {
+				branch = cvrf_branch_parse(reader);
+				if (branch)
+					oscap_list_add(tree->branches, branch);
```

Comment:
this looks like error checking but no error is reported, can branch == NULL happen here?